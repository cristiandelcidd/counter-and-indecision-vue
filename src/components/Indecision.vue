<template>
  <img v-if="image" :src="image" alt="bg" />
  <div class="bg-dark"></div>

  <div class="indecision-container">
    <input type="text" placeholder="Hazme una pregunta" v-model="question" />
    <p>Recuerda terminar con una signo de interrogación (?)</p>

    <div v-if="isValidQuestion">
      <h2>{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      question: null,
      answer: null,
      image: null,
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = "Pensando...";

        const response = await fetch("https://yesno.wtf/api");
        const { answer, image } = await response.json();

        this.answer = answer === "no" ? "No!" : "Sí!";
        this.image = image;
      } catch (error) {
        console.log("Indecision component: ", error);

        this.answer = "Ups, something went wrong!";
        this.image = null;
      }
    },
  },
  watch: {
    question(value /*, oldValue*/) {
      this.isValidQuestion = false;

      console.log({ value });

      if (!value.includes("?")) return;

      this.isValidQuestion = true;
      this.getAnswer();
    },
  },
};
</script>

<style>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  margin-top: 2rem;
  position: relative;
  z-index: 99;
}

input {
  border-radius: 5px;
  border: none;
  padding: 10px 15px;
  width: 250px;
}

input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin: 1rem 0 0;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
