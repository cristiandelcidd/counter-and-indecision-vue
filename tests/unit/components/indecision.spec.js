import { shallowMount } from "@vue/test-utils";

import Indecision from "@/components/Indecision";

describe("Indecision component", () => {
  let wrapper;
  let clgSpy;

  // Mock the fetch API
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);

    clgSpy = jest.spyOn(console, "log");

    jest.clearAllMocks();
  });

  test("should be match with the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should change the input value and shouldn't shoot anything (console.log)", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

    const input = wrapper.find("input");
    await input.setValue("Hello world");

    expect(clgSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  test("should enter the '?' symbol and execute the fecth request", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

    const input = wrapper.find("input");
    await input.setValue("Hello world?");

    expect(clgSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).toHaveBeenCalled();
  });

  test("should test the getAnswer method", async () => {
    await wrapper.vm.getAnswer();

    const image = wrapper.find("img");

    expect(image.exists()).toBeTruthy();
    expect(wrapper.vm.image).toBe("https://yesno.wtf/assets/yes/2.gif");
    expect(wrapper.vm.answer).toBe("SÃ­!");
  });

  test("should test the getAnswer method - API fails", async () => {
    fetch.mockImplementationOnce(() => Promise.reject("API is down :("));

    await wrapper.vm.getAnswer();

    const image = wrapper.find("img");

    expect(image.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("Ups, something went wrong!");
  });
});
