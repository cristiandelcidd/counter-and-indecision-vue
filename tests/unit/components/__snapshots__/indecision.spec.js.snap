// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component should be match with the snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con una signo de interrogaci√≥n (?)</p>
  <!--v-if-->
</div>
`;
